### ISSUE TITLE
ENH: Change behavior for `numpy.polynomial.polynomial.Polynomial.fit` coefficients

### ISSUE THREAD

**Issue by Lolcroc** [numpy/numpy#26401]
### Proposed new feature or change:

Currently polynomial fits work as follows
```python
x = [0, 1]
y = [0, 1]
fit = Polynomial.fit(x, y, deg=1)

print(f"Coefficients for 1-order fit on {x=}, {y=}:")
print(f"Raw: {fit.coef}")
print(f"Converted: {fit.convert().coef}")
```
which prints
```
Coefficients for 1-order fit on x=[0, 1], y=[0, 1]:
Raw: [0.5 0.5]
Converted: [-1.66533454e-16  1.00000000e+00]
```

The result for the raw `fit.coef` is unexpected. The fit is scaled with respect to its current domain, which is by default the endpoints of `x`. The `fit.convert().coef` gives the expected result, but this extra call is somewhat awkward and can easily be overlooked.

I would prefer to set `domain=[-1, 1]` as a default argument, just as it is for the `numpy.polynomial.polynomial.Polynomial` class. This creates commonality, but also avoids the unexpected outcome as in the above example.

**Comment by olcc** [comment]
I would say that the expansion on the polynomial basis is accurate over the `window` range.
This range is usually [-1,1], but for the `Laguerre` polynomial basis it is [0,1].
The `domain` represents the range of the `x` data and is meant to be mapped onto the `window` range.
Setting `domain=[-1,1]` as a default argument for the `fit()` method would increase numerical error. 

However, I agree that `P.coef` is misleading, because it does not describe the coefficients of polynomial `P` in its natural variable `x`. It rather describes the coefficients of the internal representation of `P`. Most users probably only use `P.convert().coef`. 

An improvement may be to hide the coefficients into a `P._coef` variable, and let `P.coef` return `P.convert()._coef`. Of course, this would break existing code.

**Comment by rkern** [comment]
The semantics of `.coef` can't change at this time. One could propose that there be an additional attribute that is equivalent to `.convert().coef` (either computed greedily or lazily) so that this common case is easy to access.

**Comment by olcc** [comment]
A `@property` named  `P.coeffs` could do the trick, where the property docstring would explain the difference with `P.coef`.

**Comment by janash** [comment]
I agree that this behavior is extremely confusing. I was updating some lessons this afternoon I have on fitting from `np.polyfit` to `Polynomial.fit` as directed by [the documentation](https://numpy.org/doc/stable/reference/routines.polynomials.html#quick-reference) and it was very unclear to me why the coefficients were appearing different. It has taken me several hours to figure out what was going on.

In addition to A property named `coeffs` as proposed by @olcc, I think most people would also expect the string representation (accessed through `print`) to show the unscaled fit as well.

**Comment by charris** [comment]
Note that the two-step procedure produces a more accurate result, sometimes by many orders of magnitude. The shifting and scaling should be taught in class, the naive method can lead to poor results in practice.

**Comment by janash** [comment]
@charris - Regardless of opinions on teaching shifting and scaling, the point remains that the polynomial coefficients as reported from `Polynomial.fit` will be unexpected for many users.

**Comment by charris** [comment]
> will be unexpected for many users.

But they shouldn't be, that is my point.

**Comment by ggggggggg** [comment]
> > will be unexpected for many users.
> 
> But they shouldn't be, that is my point.

You can't expect users to all have taken a class on polynomial fitting, and to remember it, and to ASSUME that the API in numpy reflects what they learned. Because ultimately there are two different both good meaning for the coefficients. Some users should expect this because `coefs` is ambiguous, it has two distinct and different meanings. It will therefore always be the case that many users expect one meaning and many expect another, no API that chooses one meaning as blessed can ever avoid breaking expectations or many users. So `coefs` should be deprecated and replaced with two properties that express intention. 
`natural_coefs` equivalent to `.convert().coef` meaning the coefficents in the natural variable 'x'
`domain_coefs` equivalent to `.coef` meaning the coefficients in the variable scaled to the domain.

I'm not tied to either name.

**Comment by olcc** [comment]
@ggggggggg: Note that the second name you suggest is wrong. Indeed, the internal range is the `window`, so the name could be `internal_coef` or `window_coef`, but not `domain_coef`.

**Comment by olcc** [comment]
Pull request: https://github.com/numpy/numpy/pull/27232

**Comment by mhvk** [comment]
I was not part of the discussion here but saw the PR in #27232. As I write there, I think adding a new property with a very similar name is a bad idea. But I also do not like the addition more generally. Since it is probably best to keep the discussion in one place, I thought I better write about that general dislike here.

Mostly, it is just echo'ing @charris comments that if you do polynomial fits, you very much risk making major rounding errors unless you scale your domain. I really like that the `Polynomial` class automatically takes care of that, and see it as a benefit that it creates some hoops to jump through before you get the unscaled coefficients - those hoops should encourage people to think what is really needed, and hopefully realize that, e.g., the proper way to evaluate the polynomial is to call it, not use the coefficients directly.

Also, at least for me, in practice there really has not been much need for the non-scaled coefficients. Indeed, for other polynomial types like `Chebyshev` or `Legendre`, would one ever want the converted coefficients? Those types really only make sense on the `[-1, 1]` interval. And even for regular polynomials, the moment you realize that your `x` and `y` usually have units, one sees that in many cases you want to rescale your values, so that your coefficients are dimensionless (not all, of course, in `s = s0 +v (t-t0) + a (t-t0)^2` the unit-full coefficients do have physical meaning).

What I do think would really help is clearer documentation. Partially on the classes themselves: `.coef` is just an attribute on the class, and something like `pol.coef?` gives the help string of `ndarray` rather than something more informative.  I think it might be good to make this a `property` (returning a private `_coef` attribute), so one can add a proper docstring. It might also help to add some usage examples to the main class docstrings. Furthermore, the main documentation page, at https://numpy.org/doc/stable/reference/routines.polynomials.html, really could use some TLC; as is, it reflects too much the state when @charris wrote everything, e.g., starting with a discussion of differences that were relevant in 2009, when numpy 1.4 was released, but not so relevant now...

**Comment by olcc** [comment]
> I really like that the Polynomial class automatically takes care of that, and see it as a benefit
> the proper way to evaluate the polynomial is to call it, not use the coefficients directly.

Sure

> at least for me, in practice there really has not been much need for the non-scaled coefficients

This is because you talk from the expert point of view :-)
Most people who had ground courses on polynomials just expect `P.coef` to return the natural coefficients.
The absence of documentation attached to that attribute creates a huge confusion for them.

> I think it might be good to make this a property (returning a private _coef attribute), so one can add a proper docstring.

This sounds nice and reasonable (but requies to modify all the occurrences of `coef` in the source code).

**Comment by mhvk** [comment]
> This sounds nice and reasonable (but requies to modify all the occurrences of `coef` in the source code).

I don't think that is too big of a problem: not changing `self.coef` to `self._coef` just gives an ever-so-slight slowdown -- I doubt it will matter for any timing. 

Of course, one can be more judicious -- or just do a quick `sed -i 's/\.coef/\._coef/g' numpy/polynomial/polybase.py` (only file where it seems to be used).

**Comment by olcc** [comment]
Adding a link to https://github.com/numpy/numpy/issues/25724

**Comment by egaudrain** [comment]
Bumped my head on this as well today. This module seems very promising (I actually love the clever conditioning), but it feels a bit half-baked: first the naming conventions do not feel very consistent, and more importantly the documentation could be improved.

If one is still wondering if `coef` and `_coef` should co-exist, I want to add that the naming conventions for this whole module could do with a serious rethink anyway. I mean, I'm told to replace `np.polyfit` with the ridiculous `np.polynomial.polynomial.Polynomial.fit`, but then integral is unnecessarily shortened as `integ`...!? At least it could be consistent with other numpy/scipy modules.

And the documentation could be improved:

- https://numpy.org/doc/stable/reference/generated/numpy.polynomial.polynomial.Polynomial.html makes no mention of the shifting-scaling representation of polynomials it is using. It gives the impression that it is just a series of coefficients. I get that internals are often hidden from the end user for simplification, but since the way the coefficients are given is directly impacted by it, this should be explicit.
- There is a mention of window and domain in https://numpy.org/doc/stable/reference/routines.polynomials.html#transition-guide, but it is just some general warning, with no explanation of what they are (give us a mathematical formula).
- On that page, there is a link to https://numpy.org/doc/stable/reference/routines.polynomials.classes promising further details on window and domain. There, it again only states that instances with different windows and/or domains do not mix. It then says "We will deal with the domain and window when we get to fitting, for the moment we ignore them..." Scrolling to "Fitting", in between two figures, there is actually some explanation of why window and domain are used, and what they are. Nice! This is where it should have pointed to from the start.
- The doc of the `Polynomial` class is not actually listing the attributes of the class, which makes the whole shifting-scaling all the more invisible. I understand that all the parameters of the constructor are also attributes, but that should be explicit, and somehow, one of the parameters (`symbol`) is also listed as attribute.
- The description of the `window` parameter makes reference to `domain`, but clicking on this link leads to the attribute's docstring, which is empty. Another missed opportunity to explain the underlying representation.
- The doc of the `fit` method keeps mentioning that it is returning a "series instance", when it is, in fact, a `Polynomial` instance. These might be the same, but that is not evident to the user and reinforces the impression that a `Polynomial` instance is just a series of coefficients. There is also some mysterious reference to domain, but without the explicit part in the `Polynomial` class help page, it just takes way too much effort to decipher. Just point to the relevant part of the convenience classes page.
- Adding some examples in the method help would also be another way of clarifying the scaling-shifting issue.

Since not everybody will get to the documentation through the transition pages (and these are meant to be temporary), I feel it would be most useful to have a description of the internal representation of polynomial series, and the role of `window` and `domain`, in the doc of the `Polynomial` class.

**Comment by mhvk** [comment]
@egaudrain - you're right that the documentation could use some work - but obviously nobody got around to it in the past decade or so....  Are you interested in contributing? It could be in small steps (and I agree best done in the documentation proper, not the transition guide; the goal would be to get rid of that!).

**Comment by olcc** [comment]
Hello,
- A proposal to go forward was agreed above (see my message on Aug 18, 2024 and the message before).
- It took several discussion before converging to this agreement (discussion here, discussion on the mailing list, refused PR). This drained my energy, and finally I have not written the new PR yet.
- I am still willing to do it.
